File: src/object/arrow.rs
use crate::object::rect::Side;

/// Direction of travel along an arrow segment.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Dir {
    Right,
    Left,
    Up,
    Down,
}

/// Unresolved arrow parsed from DSL (references object IDs).
#[derive(Debug, Clone)]
pub struct Arrow {
    pub src_id: String,
    pub src_side: Side,
    pub dst_id: String,
    pub dst_side: Side,
}

/// Resolved arrow with computed waypoints (ready for rendering).
#[derive(Debug, Clone)]
pub struct ResolvedArrow {
    /// Waypoints from source to destination (2..=5 points).
    /// First point = source anchor (outside border).
    /// Last point = destination anchor (on border).
    pub waypoints: Vec<(usize, usize)>,
    /// Custom arrowhead character (overrides global/default).
    pub head: Option<char>,
    /// Bidirectional mode: arrowhead on both ends.
    pub both: bool,
    /// Legend text displayed near the arrow.
    pub legend: Option<super::Legend>,
}

/// Converts a Side to the outgoing direction from that side.
fn side_to_outgoing_dir(side: Side) -> Dir {
    match side {
        Side::Top => Dir::Up,
        Side::Right => Dir::Right,
        Side::Bottom => Dir::Down,
        Side::Left => Dir::Left,
    }
}

/// Converts a Side to the incoming direction toward that side.
fn side_to_incoming_dir(side: Side) -> Dir {
    match side {
        Side::Top => Dir::Down,
        Side::Right => Dir::Left,
        Side::Bottom => Dir::Up,
        Side::Left => Dir::Right,
    }
}

/// Default gap for self-loop and short-distance routing.
const ROUTING_GAP: usize = 2;

/// Computes adaptive gap based on Manhattan distance between endpoints.
/// Longer distances get larger gaps to reduce crossing with other elements.
fn adaptive_gap(sx: usize, sy: usize, ex: usize, ey: usize) -> usize {
    let dist = sx.abs_diff(ex) + sy.abs_diff(ey);
    match dist {
        0..=10 => 2,
        11..=25 => 3,
        _ => 4,
    }
}

/// Computes the route waypoints between two anchor points.
///
/// - `(sx, sy)`: source anchor (1 cell outside source border)
/// - `src_side`: which side of the source rect
/// - `(ex, ey)`: destination anchor (on destination border)
/// - `dst_side`: which side of the destination rect
///
/// Returns waypoints including start and end points.
pub fn compute_route(
    sx: usize,
    sy: usize,
    src_side: Side,
    ex: usize,
    ey: usize,
    dst_side: Side,
) -> Vec<(usize, usize)> {
    let src_dir = side_to_outgoing_dir(src_side);
    let dst_dir = side_to_incoming_dir(dst_side);

    // Check if directions are perpendicular or parallel
    let src_horizontal = matches!(src_dir, Dir::Left | Dir::Right);
    let dst_horizontal = matches!(dst_dir, Dir::Left | Dir::Right);
    let perpendicular = src_horizontal != dst_horizontal;

    if perpendicular {
        // Perpendicular: try 1-bend (L-shape) or 3-bend
        route_perpendicular(sx, sy, src_dir, ex, ey, dst_dir)
    } else {
        // Parallel: try 0-bend (straight) or 2-bend (Z/U/ㄷ)
        route_parallel(sx, sy, src_dir, ex, ey, dst_dir)
    }
}

/// Routes when src and dst directions are perpendicular (1 or 3 bends).
fn route_perpendicular(
    sx: usize,
    sy: usize,
    src_dir: Dir,
    ex: usize,
    ey: usize,
    dst_dir: Dir,
) -> Vec<(usize, usize)> {
    // Try L-shape: 1 bend at the intersection of src's travel axis and dst's travel axis
    let bend = match (src_dir, dst_dir) {
        // src goes horizontal, dst goes vertical
        (Dir::Right | Dir::Left, Dir::Down | Dir::Up) => (ex, sy),
        // src goes vertical, dst goes horizontal
        (Dir::Down | Dir::Up, Dir::Right | Dir::Left) => (sx, ey),
        _ => unreachable!(),
    };

    // Check if the L-shape is "favorable" (directions align with target positions)
    if is_favorable_bend(sx, sy, src_dir, bend.0, bend.1)
        && is_favorable_bend(bend.0, bend.1, dst_dir, ex, ey)
    {
        return vec![(sx, sy), bend, (ex, ey)];
    }

    // 3-bend: need intermediate waypoints
    route_3bend(sx, sy, src_dir, ex, ey, dst_dir)
}

/// Routes when src and dst directions are parallel (0 or 2 bends).
fn route_parallel(
    sx: usize,
    sy: usize,
    src_dir: Dir,
    ex: usize,
    ey: usize,
    dst_dir: Dir,
) -> Vec<(usize, usize)> {
    let horizontal = matches!(src_dir, Dir::Left | Dir::Right);
    let same_direction = src_dir == dst_dir;

    if same_direction {
        // Same direction = natural flow: try straight, then Z-shape
        let is_straight = if horizontal {
            sy == ey
                && ((src_dir == Dir::Right && ex >= sx) || (src_dir == Dir::Left && ex <= sx))
        } else {
            sx == ex
                && ((src_dir == Dir::Down && ey >= sy) || (src_dir == Dir::Up && ey <= sy))
        };
        if is_straight {
            return vec![(sx, sy), (ex, ey)];
        }
    }

    // 2-bend (Z or ㄷ shape)
    // same_direction but not aligned → Z-shape (midpoint between)
    // opposite directions → ㄷ-shape (detour past both endpoints)
    route_2bend(sx, sy, src_dir, ex, ey, dst_dir, horizontal, same_direction)
}

/// Checks if traveling from (fx,fy) in direction `dir` can reach (tx,ty).
fn is_favorable_bend(fx: usize, fy: usize, dir: Dir, tx: usize, ty: usize) -> bool {
    match dir {
        Dir::Right => tx >= fx,
        Dir::Left => tx <= fx,
        Dir::Down => ty >= fy,
        Dir::Up => ty <= fy,
    }
}

/// 2-bend routing for parallel directions.
#[allow(clippy::too_many_arguments)]
fn route_2bend(
    sx: usize,
    sy: usize,
    src_dir: Dir,
    ex: usize,
    ey: usize,
    _dst_dir: Dir,
    horizontal: bool,
    same_direction: bool,
) -> Vec<(usize, usize)> {
    if horizontal {
        if same_direction {
            let mid_x = midpoint(sx, ex);
            vec![(sx, sy), (mid_x, sy), (mid_x, ey), (ex, ey)]
        } else {
            let gap = adaptive_gap(sx, sy, ex, ey);
            let mid_x = match src_dir {
                Dir::Right => sx.max(ex) + gap,
                Dir::Left => sx.min(ex).saturating_sub(gap),
                _ => unreachable!(),
            };
            vec![(sx, sy), (mid_x, sy), (mid_x, ey), (ex, ey)]
        }
    } else if same_direction {
        let mid_y = midpoint(sy, ey);
        vec![(sx, sy), (sx, mid_y), (ex, mid_y), (ex, ey)]
    } else {
        let gap = adaptive_gap(sx, sy, ex, ey);
        let mid_y = match src_dir {
            Dir::Down => sy.max(ey) + gap,
            Dir::Up => sy.min(ey).saturating_sub(gap),
            _ => unreachable!(),
        };
        vec![(sx, sy), (sx, mid_y), (ex, mid_y), (ex, ey)]
    }
}

/// 3-bend routing for perpendicular directions that can't form a simple L.
fn route_3bend(
    sx: usize,
    sy: usize,
    src_dir: Dir,
    ex: usize,
    ey: usize,
    _dst_dir: Dir,
) -> Vec<(usize, usize)> {
    // Create a detour: extend src direction, then perpendicular, then into dst
    let horizontal_src = matches!(src_dir, Dir::Left | Dir::Right);

    if horizontal_src {
        // src goes horizontal, dst goes vertical (but unfavorable L)
        // Extend horizontal to midpoint x, then vertical to ey, then horizontal to ex
        let mid_x = midpoint(sx, ex);
        vec![(sx, sy), (mid_x, sy), (mid_x, ey), (ex, ey)]
    } else {
        // src goes vertical, dst goes horizontal (but unfavorable L)
        let mid_y = midpoint(sy, ey);
        vec![(sx, sy), (sx, mid_y), (ex, mid_y), (ex, ey)]
    }
}

/// Simple midpoint calculation.
fn midpoint(a: usize, b: usize) -> usize {
    (a + b) / 2
}

/// Computes a self-loop route (src and dst on the same object).
/// Always routes *outside* the object using a ㄷ-shape detour.
pub fn compute_self_loop(
    sx: usize,
    sy: usize,
    src_side: Side,
    ex: usize,
    ey: usize,
    dst_side: Side,
) -> Vec<(usize, usize)> {
    let gap = ROUTING_GAP;
    let src_dir = side_to_outgoing_dir(src_side);
    // For self-loop, extend dst clearance point *away* from the object (outgoing direction)
    let dst_away = side_to_outgoing_dir(dst_side);

    // Extend from src anchor in src_dir by gap, then route to dst approach, then to dst
    let (p1c, p1r) = extend_point(sx, sy, src_dir, gap);
    let (p2c, p2r) = extend_point(ex, ey, dst_away, gap);

    // If p1 and p2 share an axis, we can do a simple 2-bend
    if p1c == p2c || p1r == p2r {
        vec![(sx, sy), (p1c, p1r), (p2c, p2r), (ex, ey)]
    } else {
        // 3-bend: connect via an intermediate segment
        // Choose the corner path based on src/dst directions
        let src_horizontal = matches!(src_dir, Dir::Left | Dir::Right);
        if src_horizontal {
            // p1 is on horizontal extension, p2 on vertical extension (or vice versa)
            vec![(sx, sy), (p1c, p1r), (p1c, p2r), (p2c, p2r), (ex, ey)]
        } else {
            vec![(sx, sy), (p1c, p1r), (p2c, p1r), (p2c, p2r), (ex, ey)]
        }
    }
}

/// Extends a point in the given direction by `distance`.
fn extend_point(c: usize, r: usize, dir: Dir, distance: usize) -> (usize, usize) {
    match dir {
        Dir::Right => (c + distance, r),
        Dir::Left => (c.saturating_sub(distance), r),
        Dir::Down => (c, r + distance),
        Dir::Up => (c, r.saturating_sub(distance)),
    }
}

/// Returns the corner character for a waypoint where incoming direction meets outgoing direction.
pub fn corner_char(incoming: Dir, outgoing: Dir) -> char {
    match (incoming, outgoing) {
        (Dir::Right, Dir::Down) => '┐',
        (Dir::Right, Dir::Up) => '┘',
        (Dir::Left, Dir::Down) => '┌',
        (Dir::Left, Dir::Up) => '└',
        (Dir::Down, Dir::Right) => '└',
        (Dir::Down, Dir::Left) => '┘',
        (Dir::Up, Dir::Right) => '┌',
        (Dir::Up, Dir::Left) => '┐',
        (Dir::Left, Dir::Left) | (Dir::Right, Dir::Right) => '─',
        (Dir::Up, Dir::Up) | (Dir::Down, Dir::Down) => '│',
        _ => '┼',
    }
}

/// Computes the direction from point (fx,fy) to point (tx,ty).
/// Points must share an axis (same row or same column).
pub fn segment_dir(fx: usize, fy: usize, tx: usize, ty: usize) -> Dir {
    if fy == ty {
        if tx > fx { Dir::Right } else { Dir::Left }
    } else if ty > fy {
        Dir::Down
    } else {
        Dir::Up
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn straight_horizontal_route() {
        // right.src → left.dst, aligned
        let wp = compute_route(10, 5, Side::Right, 20, 5, Side::Left);
        assert_eq!(wp, vec![(10, 5), (20, 5)]);
    }

    #[test]
    fn straight_vertical_route() {
        let wp = compute_route(5, 3, Side::Bottom, 5, 10, Side::Top);
        assert_eq!(wp, vec![(5, 3), (5, 10)]);
    }

    #[test]
    fn l_shape_right_to_top() {
        // src exits right, dst enters from top (down)
        let wp = compute_route(10, 5, Side::Right, 20, 10, Side::Top);
        assert_eq!(wp, vec![(10, 5), (20, 5), (20, 10)]);
    }

    #[test]
    fn l_shape_bottom_to_left() {
        let wp = compute_route(5, 5, Side::Bottom, 15, 10, Side::Left);
        assert_eq!(wp, vec![(5, 5), (5, 10), (15, 10)]);
    }

    #[test]
    fn z_shape_opposite_vertical() {
        // Both vertical, opposite directions, not aligned
        let wp = compute_route(5, 3, Side::Bottom, 15, 10, Side::Top);
        // midpoint y = (3+10)/2 = 6
        assert_eq!(wp, vec![(5, 3), (5, 6), (15, 6), (15, 10)]);
    }

    #[test]
    fn u_shape_same_side_right() {
        // Both exit/enter from right → ㄷ-shape
        let wp = compute_route(10, 2, Side::Right, 10, 8, Side::Right);
        // mid_x = max(10,10) + 2 = 12
        assert_eq!(wp, vec![(10, 2), (12, 2), (12, 8), (10, 8)]);
    }

    #[test]
    fn corner_chars_correct() {
        assert_eq!(corner_char(Dir::Right, Dir::Down), '┐');
        assert_eq!(corner_char(Dir::Right, Dir::Up), '┘');
        assert_eq!(corner_char(Dir::Down, Dir::Right), '└');
        assert_eq!(corner_char(Dir::Up, Dir::Right), '┌');
    }
}

File: internal/object/arrow.go
package object

import "github.com/silee9019/unicode-diagram/internal/width"

type Dir int

const (
	DirRight Dir = iota
	DirLeft
	DirUp
	DirDown
)

var arrowheadFamilies = [][4]rune{
	{'▶', '◀', '▲', '▼'}, // default (filled triangle)
	{'→', '←', '↑', '↓'}, // arrow
	{'▷', '◁', '△', '▽'}, // outline triangle
	{'⇒', '⇐', '⇑', '⇓'}, // double arrow
}

func IsValidArrowhead(ch rune) bool {
	for _, f := range arrowheadFamilies {
		for _, c := range f {
			if c == ch {
				return true
			}
		}
	}
	return false
}

func ValidArrowheadChars() []rune {
	var chars []rune
	for _, f := range arrowheadFamilies {
		chars = append(chars, f[:]...)
	}
	return chars
}

func ResolveArrowhead(ch rune, dir Dir) rune {
	idx := dirIndex(dir)
	for _, f := range arrowheadFamilies {
		for _, c := range f {
			if c == ch {
				return f[idx]
			}
		}
	}
	return arrowheadFamilies[0][idx]
}

func DefaultArrowhead(dir Dir) rune {
	return arrowheadFamilies[0][dirIndex(dir)]
}

func dirIndex(dir Dir) int {
	switch dir {
	case DirRight:
		return 0
	case DirLeft:
		return 1
	case DirUp:
		return 2
	case DirDown:
		return 3
	}
	return 0
}

type ResolvedArrow struct {
	Waypoints [][2]int // (col, row) pairs
	Head      rune
	HasHead   bool
	Both      bool
	Legend    *Legend
}

func SideToOutgoingDir(side Side) Dir {
	switch side {
	case SideTop:
		return DirUp
	case SideRight:
		return DirRight
	case SideBottom:
		return DirDown
	case SideLeft:
		return DirLeft
	}
	return DirRight
}

func SideToIncomingDir(side Side) Dir {
	switch side {
	case SideTop:
		return DirDown
	case SideRight:
		return DirLeft
	case SideBottom:
		return DirUp
	case SideLeft:
		return DirRight
	}
	return DirRight
}

const routingGap = 2

func adaptiveGap(sx, sy, ex, ey int) int {
	dist := abs(sx-ex) + abs(sy-ey)
	switch {
	case dist <= 10:
		return 2
	case dist <= 25:
		return 3
	default:
		return 4
	}
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func ComputeRoute(sx, sy int, srcSide Side, ex, ey int, dstSide Side) [][2]int {
	srcDir := SideToOutgoingDir(srcSide)
	dstDir := SideToIncomingDir(dstSide)

	srcHorizontal := srcDir == DirLeft || srcDir == DirRight
	dstHorizontal := dstDir == DirLeft || dstDir == DirRight
	perpendicular := srcHorizontal != dstHorizontal

	if perpendicular {
		return routePerpendicular(sx, sy, srcDir, ex, ey, dstDir)
	}
	return routeParallel(sx, sy, srcDir, ex, ey, dstDir)
}

func routePerpendicular(sx, sy int, srcDir Dir, ex, ey int, dstDir Dir) [][2]int {
	var bend [2]int
	switch {
	case (srcDir == DirRight || srcDir == DirLeft) && (dstDir == DirDown || dstDir == DirUp):
		bend = [2]int{ex, sy}
	case (srcDir == DirDown || srcDir == DirUp) && (dstDir == DirRight || dstDir == DirLeft):
		bend = [2]int{sx, ey}
	}

	if isFavorableBend(sx, sy, srcDir, bend[0], bend[1]) &&
		isFavorableBend(bend[0], bend[1], dstDir, ex, ey) {
		return [][2]int{{sx, sy}, bend, {ex, ey}}
	}

	return route3Bend(sx, sy, srcDir, ex, ey)
}

func routeParallel(sx, sy int, srcDir Dir, ex, ey int, dstDir Dir) [][2]int {
	horizontal := srcDir == DirLeft || srcDir == DirRight
	sameDirection := srcDir == dstDir

	if sameDirection {
		var isStraight bool
		if horizontal {
			isStraight = sy == ey &&
				((srcDir == DirRight && ex >= sx) || (srcDir == DirLeft && ex <= sx))
		} else {
			isStraight = sx == ex &&
				((srcDir == DirDown && ey >= sy) || (srcDir == DirUp && ey <= sy))
		}
		if isStraight {
			return [][2]int{{sx, sy}, {ex, ey}}
		}
	}

	return route2Bend(sx, sy, srcDir, ex, ey, horizontal, sameDirection)
}

func isFavorableBend(fx, fy int, dir Dir, tx, ty int) bool {
	switch dir {
	case DirRight:
		return tx >= fx
	case DirLeft:
		return tx <= fx
	case DirDown:
		return ty >= fy
	case DirUp:
		return ty <= fy
	}
	return false
}

func route2Bend(sx, sy int, srcDir Dir, ex, ey int, horizontal, sameDirection bool) [][2]int {
	if horizontal {
		if sameDirection {
			midX := (sx + ex) / 2
			return [][2]int{{sx, sy}, {midX, sy}, {midX, ey}, {ex, ey}}
		}
		gap := adaptiveGap(sx, sy, ex, ey)
		var midX int
		if srcDir == DirRight {
			midX = max(sx, ex) + gap
		} else {
			midX = max(min(sx, ex)-gap, 0)
		}
		return [][2]int{{sx, sy}, {midX, sy}, {midX, ey}, {ex, ey}}
	}

	if sameDirection {
		midY := (sy + ey) / 2
		return [][2]int{{sx, sy}, {sx, midY}, {ex, midY}, {ex, ey}}
	}
	gap := adaptiveGap(sx, sy, ex, ey)
	var midY int
	if srcDir == DirDown {
		midY = max(sy, ey) + gap
	} else {
		midY = max(min(sy, ey)-gap, 0)
	}
	return [][2]int{{sx, sy}, {sx, midY}, {ex, midY}, {ex, ey}}
}

func route3Bend(sx, sy int, srcDir Dir, ex, ey int) [][2]int {
	horizontalSrc := srcDir == DirLeft || srcDir == DirRight
	if horizontalSrc {
		midX := (sx + ex) / 2
		return [][2]int{{sx, sy}, {midX, sy}, {midX, ey}, {ex, ey}}
	}
	midY := (sy + ey) / 2
	return [][2]int{{sx, sy}, {sx, midY}, {ex, midY}, {ex, ey}}
}

func ComputeSelfLoop(sx, sy int, srcSide Side, ex, ey int, dstSide Side) [][2]int {
	gap := routingGap
	srcDir := SideToOutgoingDir(srcSide)
	dstAway := SideToOutgoingDir(dstSide)

	p1c, p1r := extendPoint(sx, sy, srcDir, gap)
	p2c, p2r := extendPoint(ex, ey, dstAway, gap)

	var raw [][2]int
	if p1c == p2c || p1r == p2r {
		raw = [][2]int{{sx, sy}, {p1c, p1r}, {p2c, p2r}, {ex, ey}}
	} else {
		srcHorizontal := srcDir == DirLeft || srcDir == DirRight
		if srcHorizontal {
			raw = [][2]int{{sx, sy}, {p1c, p1r}, {p1c, p2r}, {p2c, p2r}, {ex, ey}}
		} else {
			raw = [][2]int{{sx, sy}, {p1c, p1r}, {p2c, p1r}, {p2c, p2r}, {ex, ey}}
		}
	}

	// Deduplicate consecutive points
	deduped := make([][2]int, 0, len(raw))
	for _, pt := range raw {
		if len(deduped) == 0 || deduped[len(deduped)-1] != pt {
			deduped = append(deduped, pt)
		}
	}
	return deduped
}

func extendPoint(c, r int, dir Dir, distance int) (int, int) {
	switch dir {
	case DirRight:
		return c + distance, r
	case DirLeft:
		return max(c-distance, 0), r
	case DirDown:
		return c, r + distance
	case DirUp:
		return c, max(r-distance, 0)
	}
	return c, r
}

func CornerChar(incoming, outgoing Dir) rune {
	switch {
	case incoming == DirRight && outgoing == DirDown:
		return '┐'
	case incoming == DirRight && outgoing == DirUp:
		return '┘'
	case incoming == DirLeft && outgoing == DirDown:
		return '┌'
	case incoming == DirLeft && outgoing == DirUp:
		return '└'
	case incoming == DirDown && outgoing == DirRight:
		return '└'
	case incoming == DirDown && outgoing == DirLeft:
		return '┘'
	case incoming == DirUp && outgoing == DirRight:
		return '┌'
	case incoming == DirUp && outgoing == DirLeft:
		return '┐'
	case (incoming == DirLeft && outgoing == DirLeft) || (incoming == DirRight && outgoing == DirRight):
		return '─'
	case (incoming == DirUp && outgoing == DirUp) || (incoming == DirDown && outgoing == DirDown):
		return '│'
	default:
		return '┼'
	}
}

// LegendPosition computes arrow legend position and column based on waypoints, legend config, and align.
// Returns (col, row, textWidth).
func LegendPosition(wp [][2]int, legend *Legend) (int, int, int) {
	if len(wp) < 2 {
		return 0, 0, 0
	}

	// For bent arrows, pick the longest segment from the second one onwards.
	segIdx := 0
	if len(wp) >= 3 {
		bestLen := 0
		for i := 1; i < len(wp)-1; i++ {
			l := abs(wp[i][0]-wp[i+1][0]) + abs(wp[i][1]-wp[i+1][1])
			if l >= bestLen {
				bestLen = l
				segIdx = i
			}
		}
	}

	fc, fr := wp[segIdx][0], wp[segIdx][1]
	tc, tr := wp[segIdx+1][0], wp[segIdx+1][1]
	midC := (fc + tc) / 2
	midR := (fr + tr) / 2
	dir := SegmentDir(fc, fr, tc, tr)
	isHorizontal := dir == DirLeft || dir == DirRight

	effectivePos := legend.Pos
	if effectivePos == LegendAuto {
		if isHorizontal {
			effectivePos = LegendTop
		} else {
			effectivePos = LegendRight
		}
	}

	textW := width.StrWidth(legend.Text)

	var lgCol, lgRow int
	switch effectivePos {
	case LegendTop, LegendBottom:
		switch legend.Align {
		case AlignLeft:
			lgCol = midC
		case AlignCenter:
			lgCol = max(midC-textW/2, 0)
		case AlignRight:
			lgCol = max(midC-textW, 0)
		}
		if effectivePos == LegendTop {
			lgRow = max(midR, 1) - 1
		} else {
			lgRow = midR + 1
		}
	case LegendLeft:
		lgCol = max(midC-textW-1, 0)
		lgRow = midR
	case LegendRight:
		lgCol = midC + 1
		lgRow = midR
	}

	return lgCol, lgRow, textW
}

func SegmentDir(fx, fy, tx, ty int) Dir {
	if fy == ty {
		if tx > fx {
			return DirRight
		}
		return DirLeft
	}
	if ty > fy {
		return DirDown
	}
	return DirUp
}
